### DISCUSSION TITLE
Need assistance with my simple pipeline.

### DISCUSSION THREAD

**Discussion by prasmish** [halide/Halide#7656] [Q&A] [answered]
Hi everyone,
I was testing a very small pipeline as follows:
```
    ImageParam input(type_of<int32_t>(), 1);
    
    Func f, g, output;
    Var x, y;
    int vector_size = 128;
    
    //Pipeline
    f(x) = input(x);
    g(x) = f(x);
    output(x,y) = f(x) + g(x);

    //Schedule of the pipeline
    f.compute_at(output, y).store_at(output, Var::outermost()).vectorize(x, vector_size);
    g.compute_at(output, y).store_at(output, Var::outermost()).vectorize(x, vector_size);
```
Running this pipeline on input buffer of size 1024 ,output buffer of size 1024x1024, and vector_size = 128; it throws an error saying:
```
Error: Input buffer p0 is accessed at -128, which is before the min (0) in dimension 0
Aborted
```
On further diagnosing, I found that by bypassing the SlidingWindow optimization pass, the pipeline successfully runs. I suspect that SlidingWindow optimization pass is manipulating the minimum of buffers in such a way that it is leading to this error. The above pipeline also runs fine when I use the `vector_size = 1`.
Can anyone help me in this?
 
@abadams

**Comment by abadams** [comment, marked as answer]
Here's what's going on:

Sliding window is detecting that f and g only need to be computed for the first value of y, and can be reused thereafter. For all scanlines but the first, the required extent of f and g is zero. More precisely, if it has inferred a min coordinate of 0 and a max coordinate of -1. Bounds inference is then looking at that -1 for the max coordinate, combined with the vectorization with a ShiftInwards tail strategy (the default), and deciding that for the last vector of work it's going to compute coordinates from [-128...-1] inclusive, and is therefore going to read from before the beginning of the input.

I'll see what I need to do to convince bounds inference that there are actually zero vectors worth of work going on, so it needs to chill. In the meantime, an equivalent schedule is just compute-rooting f and g.

**Reply by prasmish** [reply]
Thanks a lot for the detailed explanation of what is happening under the hood.
